addons/mail/models/discuss/discuss_channel.py

```diff
@@ -715,18 +715,20 @@ def _find_or_create_persona_for_channel(self, guest_name, timezone, country_code
             return member.partner_id, member.guest_id
         if not self.env.user._is_public():
             self.add_members([self.env.user.partner_id.id], post_joined_message=post_joined_message)
-        elif self.env.user._is_public():
-            is_guest_known = self.env["mail.guest"]._get_guest_from_context()
-            country_id = self.env["res.country"].search([("code", "=", country_code)]).id
-            guest = self.env["mail.guest"]._find_or_create_for_channel(
-                channel=self,
-                country_id=country_id,
-                name=guest_name,
-                post_joined_message=post_joined_message,
-                timezone=timezone,
-            )
-            if not is_guest_known:
+        else:
+            guest = self.env["mail.guest"]._get_guest_from_context()
+            if not guest:
+                guest = self.env["mail.guest"].create(
+                    {
+                        "country_id": self.env["res.country"].search([("code", "=", country_code)]).id,
+                        "lang": get_lang(self.env).code,
+                        "name": guest_name,
+                        "timezone": timezone,
+                    }
+                ).sudo(False)
                 guest._set_auth_cookie()
+                self = self.with_context(guest=guest)
+            self.add_members(guest_ids=guest.ids, post_joined_message=post_joined_message)
         return self.env.user.partner_id if not guest else None, guest
```

Comment:
isn't guest necessarily set at this step?

Reply:
Note: it's better to return an empty partner recordset rather than None for api consistency

Reply:
updated to return a recordset instead of `None`. Guest could be `None` but was updated as well to return an empty recordset instead.